### PR TITLE
cc-wrapper: More intelligent sierra hack

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

I've done this twice before on stable branches https://github.com/NixOS/nixpkgs/pull/29096 https://github.com/NixOS/nixpkgs/pull/29935. This begs the question, why on earth did I not do it on master? The reason is `ld-wrapper.sh`'s `--rpath` injection hard-coded `.so`, which I sed-ed to `.dylib`, but wanted to make handle both as some broken build systems cause the use of `.so` on Darwin. Turns out, we never inject `--rpath` on Darwin, and really shouldn't on Linux either, so let's merge!

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/build-support/bintools-wrapper/macos-sierra-reexport-hack.bash [hunk 1]
```diff
@@ -2,107 +2,245 @@
...
+    case "$1" in
+        -l)
+            echo "cctools LD does not support '-l foo'" >&2
+            exit 1
+            ;;
+        -lazy_library | -reexport_library | -upward_library | -weak_library)
+            overflowCount+=1
+            shift 2
+            ;;
+        -l* | *.so | *.dylib | -lazy-l* | -reexport-l* | -upward-l* | -weak-l*)
```
Comment (LnL7) [inline pkgs/build-support/bintools-wrapper/macos-sierra-reexport-hack.bash]: Something like `lifboo.so.2` is also pretty common.

**Comment by Ericson2314** [comment]
@copumpkin Yeah we did that for the response file stuff in the end. This is right now twice as complex as it needs to be because I accidentally committed the same script in `cc-wrapper`, so let me start there.

File: pkgs/build-support/bintools-wrapper/macos-sierra-reexport-hack.bash (hunk 1 above)
Comment (Ericson2314) [reply pkgs/build-support/bintools-wrapper/macos-sierra-reexport-hack.bash]: Oh true.

**Comment by Ericson2314** [comment]
Ok tested locally.

**Pull request merged**